R/nested.R

```diff
@@ -46,13 +49,17 @@ compute_parse_data_nested <- function(text) {
 #'   This is done recursively until we are only left with a nested tibble that
 #'   contains one row: The nested parse table.
 #' @param pd_flat A flat parse table including both terminals and non-terminals.
+#' @param top_level_expr The ids of the expressions that cannot be attributed
+#'   to another expression since they are not nested in any other expression.
+#'   This is important for terminating the recursion, i.e. further nesting is
+#'   impossible if our nested parse table has `length(top_level_expr)` rows.
 #' @seealso [compute_parse_data_nested()]
 #' @return A nested parse table.
-nest_parse_data <- function(pd_flat) {
-  if (nrow(pd_flat) <= 1) return(pd_flat)
+nest_parse_data <- function(pd_flat, top_level_expr) {
+  if (nrow(pd_flat) <= length(top_level_expr)) return(pd_flat)
```

Comment:
I think we cant' directly because the stopping condition is only half of the story. The other reason we need the ids of the top_level expressions is because we nest `pd_flat` into `internal` and `child`. 
```r
  split <-
    pd_flat %>%
    mutate_(internal = ~id %in% c(parent, top_level_expr)) %>%
    nest_("data", names(pd_flat))

```
However, we could get rid of it by also changing `mutate_()`
```r
nest_parse_data <- function(pd_flat) {
  if (all(pd_flat$parent == 0)) return(pd_flat)
  split <-
    pd_flat %>%
    mutate_(internal = ~id %in% parent | parent == 0) %>%
    nest_("data", names(pd_flat))
```
Should I do that?